### PR TITLE
merge from upstream (includes [MS-427])

### PR DESCRIPTION
**Pull request by fthuin**

### CODE REVIEW

File: src/main/kotlin/org/taktik/freehealth/middleware/dto/efact/InvoiceSender.kt [hunk 1]
```diff
@@ -30,6 +32,9 @@ class InvoiceSender {
     var phoneNumber: Long? = null
     var conventionCode: Int? = null
 
+    var isMedicalHouse:Boolean = false
+        get() = nihii != null && Arrays.asList("111", "110", "100", "101", "001", "010", "011").contains(nihii!!.toString().substring(8))
```
Comment (phil1618) [inline src/main/kotlin/org/taktik/freehealth/middleware/dto/efact/InvoiceSender.kt]: This is incorrect: 

001: Dentiste généraliste 
010: Dentiste généraliste - en formation
011: médecin spécialiste en formation en anesthésie-réanimation
100: médecin spécialiste en anesthésie-réanimation
101: médecin spécialiste en anesthésie-réanimation et médecine générale
110: ?
111: Surement maison médicale... (Exemple, Maison Médicale d'Anderlecht asbl Numéro INAMI : 8.28502.72.111)

File: src/main/kotlin/org/taktik/freehealth/middleware/format/efact/BelgianInsuranceInvoicingFormatWriter.kt [hunk 2]
```diff
@@ -59,12 +59,24 @@ class BelgianInsuranceInvoicingFormatWriter(private val writer: Writer) {
...
     }
 
-    fun getDestCode(affCode: String, invoiceSender: InvoiceSender): String {
+    fun getDestCode(affCode: String, invoiceSender: InvoiceSender, returnAffCodeIfMH: Boolean = false): String {
         val firstCode = affCode.substring(0, 3).replace("[^0-9]".toRegex(), "")
-        if (affCode.startsWith("3")) {
-            return if (Arrays.asList("305", "315", "317", "319", "323", "325").contains(firstCode)) if (invoiceSender.isSpecialist) "317" else "319" else firstCode
-        } else if (affCode.startsWith("4")) {
-            return "400"
+        if(invoiceSender.isMedicalHouse && returnAffCodeIfMH){
```
Comment (phil1618) [inline src/main/kotlin/org/taktik/freehealth/middleware/format/efact/BelgianInsuranceInvoicingFormatWriter.kt]: The MH piece of code could be isolated in its own branch.

File: src/main/kotlin/org/taktik/freehealth/middleware/format/efact/segments/Record10Description.kt [hunk 3]
```diff
@@ -29,56 +29,56 @@ object Record10Description : RecordOrSegmentDescription() {
...
+        pos = register(ZONE_DESCRIPTIONS_BY_ZONE, "11", "Reserve", null, "N", pos, 1)
+        pos = register(ZONE_DESCRIPTIONS_BY_ZONE, "12", "Reserve", null, "N", pos, 1)
+        pos = register(ZONE_DESCRIPTIONS_BY_ZONE, "13", "ContenuDeLaFacturation", null, "N", pos, 3)
+        pos = register(ZONE_DESCRIPTIONS_BY_ZONE, "14", "NumeroTiersPayant", "thirdPartyNumber", "N", pos, 12)
+        pos = register(ZONE_DESCRIPTIONS_BY_ZONE, "15", "NumeroDaccreditationCin", "accreditationCinNumber", "N", pos, 12)
+        pos = register(ZONE_DESCRIPTIONS_BY_ZONE, "16", "Reserve", null, "N", pos, 1)
+        pos = register(ZONE_DESCRIPTIONS_BY_ZONE, "17", "Reserve", null, "N", pos, 4)
+        pos = register(ZONE_DESCRIPTIONS_BY_ZONE, "18", "Reserve", null, "N", pos, 3)
+        pos = register(ZONE_DESCRIPTIONS_BY_ZONE, "19", "Reserve", null, "N", pos, 12)
+        pos = register(ZONE_DESCRIPTIONS_BY_ZONE, "20", "Reserve", null, "N", pos, 7)
```
Comment (phil1618) [inline src/main/kotlin/org/taktik/freehealth/middleware/format/efact/segments/Record10Description.kt:52]: We should check that the default value is not changed.

File: src/main/kotlin/org/taktik/freehealth/middleware/service/HubService.kt [hunk 4]
```diff
@@ -138,7 +140,7 @@ interface HubService {
         from: Instant?,
         to: Instant?,
         hubPackageId: String?
-    ): List<TherapeuticLink>
+    ): List<TherapeuticLinkMessage>
```
Comment (phil1618) [inline src/main/kotlin/org/taktik/freehealth/middleware/service/HubService.kt:143]: To be check when getting Hub therlink @nicolas-b12

File: src/main/kotlin/org/taktik/freehealth/middleware/service/impl/EfactServiceImpl.kt [hunk 5]
```diff
@@ -81,93 +81,117 @@ class EfactServiceImpl(private val stsService: STSService, private val mapper: M
...
-                    var recordAmount = 0L
-                    var recordFee = 0L
-                    var recordSup = 0L
-                    rn =
-                        iv.writeRecordHeader(rn, batch.sender!!, invoice.invoiceNumber!!, invoice.reason!!, invoice.invoiceRef!!, invoice.patient!!, invoice.ioCode!!, invoice.ignorePrescriptionDate, invoice.hospitalisedPatient, invoice.creditNote, invoice.relatedBatchSendNumber, invoice.relatedBatchYearMonth, invoice.relatedInvoiceIoCode, invoice.relatedInvoiceNumber)
-                    recordsCountPerOA[0]++
-                    recordsCount++
-                    for (it in invoice.items) {
-                        rn = iv.writeRecordContent(rn, batch.sender!!, batch.invoicingYear, batch.invoicingMonth, invoice.patient!!, invoice.ioCode!!, it)
+    override fun makeFlatFileCoreWithMetadata(batch: InvoicesBatch, isTest: Boolean): FlatFileWithMetadata {
```
Comment (phil1618) [inline src/main/kotlin/org/taktik/freehealth/middleware/service/impl/EfactServiceImpl.kt:98]: We should read the corresponding file to be sure

File: src/main/kotlin/org/taktik/freehealth/middleware/service/impl/HubServiceImpl.kt [hunk 6]
```diff
@@ -308,13 +314,36 @@ class HubServiceImpl(val stsService: STSService, val mapper: MapperFacade) : Hub
...
-            it.therapeuticlinks.map { mapper.map(it, TherapeuticLink::class.java) }
-        } ?: listOf()
+        val errors = therapeuticLinkResponse.acknowledge.errors.map {
+            Error().apply {
+                this.url = it.url;
+                this.descr = it.description.getValue();
+            }
+        };
+        val isComplete = therapeuticLinkResponse.acknowledge.isIscomplete();
+        return therapeuticLinkResponse.therapeuticlinklist.therapeuticlinks?.map {
```
Comment (phil1618) [inline src/main/kotlin/org/taktik/freehealth/middleware/service/impl/HubServiceImpl.kt]: This might have a small impact => to be checked @nicolas-b12

**Review by phil1618 — CHANGES_REQUESTED** [review]
To be check

File: src/main/kotlin/org/taktik/freehealth/middleware/service/HubService.kt (hunk 4 above)
Comment (nicolas-b12) [reply src/main/kotlin/org/taktik/freehealth/middleware/service/HubService.kt:143]: I agree with this.

File: src/main/kotlin/org/taktik/freehealth/middleware/service/impl/HubServiceImpl.kt (hunk 6 above)
Comment (nicolas-b12) [reply src/main/kotlin/org/taktik/freehealth/middleware/service/impl/HubServiceImpl.kt]: I agree with this code 👍

**Pull request merged**